src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java

```diff
@@ -0,0 +1,90 @@
+            ProgrammingAssessmentService programmingAssessmentService, ProgrammingSubmissionService programmingSubmissionService) {
+        super(authCheckService, userService);
+        this.programmingExerciseService = programmingExerciseService;
+        this.programmingAssessmentService = programmingAssessmentService;
+        this.programmingSubmissionService = programmingSubmissionService;
+    }
+
+    /**
+     * Update an assessment after a complaint was accepted.
+     *
+     * @param submissionId the id of the submission for which the assessment should be updated
+     * @param assessmentUpdate the programing assessment update
+     * @return the updated result
+     */
+    @ResponseStatus(HttpStatus.OK)
+    @PutMapping("/programming-submissions/{submissionId}/assessment-after-complaint")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Result> updateProgrammingManualResultAfterComplaint(@RequestBody ProgrammingAssessmentUpdate assessmentUpdate, @PathVariable Long submissionId) {
+        log.debug("REST request to update the assessment of manual result for submission {} after complaint.", submissionId);
+        User user = userService.getUserWithGroupsAndAuthorities();
+        ProgrammingSubmission programmingSubmission = programmingSubmissionService.findOneWithEagerResultAndFeedback(submissionId);
+        Participation participation = programmingSubmission.getParticipation();
+        long exerciseId = participation.getExercise().getId();
+        ProgrammingExercise programmingExercise = programmingExerciseService.findOne(exerciseId);
+        checkAuthorization(programmingExercise, user);
+        if (!areManualResultsAllowed(programmingExercise)) {
+            return forbidden();
+        }
+
+        Result result = programmingAssessmentService.updateAssessmentAfterComplaint(programmingSubmission.getResult(), programmingExercise, assessmentUpdate);
+        // remove circular dependencies if the results of the participation are there
+        if (result.getParticipation() != null && Hibernate.isInitialized(result.getParticipation().getResults()) && result.getParticipation().getResults() != null) {
+            result.getParticipation().setResults(null);
+        }
+
+        if (result.getParticipation() != null && result.getParticipation() instanceof StudentParticipation
+                && !authCheckService.isAtLeastInstructorForExercise(programmingExercise, user)) {
+            ((StudentParticipation) result.getParticipation()).filterSensitiveInformation();
+        }
+
+        return ResponseEntity.ok(result);
+    }
+
+    @Override
+    String getEntityName() {
+        return ENTITY_NAME;
+    }
+
+    private boolean areManualResultsAllowed(final Exercise exerciseToBeChecked) {
+        // Only allow manual results for programming exercises if option was enabled and due dates have passed
```

Comment:
I think this method already exists somewhere else. Please reuse code and do not simply copy it!

Reply:
Yes it is used inside the `ResultResource`. I would like to move it into the `ProgrammingAssessmentService`, but I don't want to inject it this service in the `ResultResource`. 
We actually need to remove all the methods specific for programming exercise from the `ResultResource` to this file, then we can easily reuse this method

Reply:
Moved it to `ProgrammingExercise` domain